augur/io/metadata.py

```diff
@@ -171,3 +172,60 @@ def read_metadata_to_dict(metadata, duplicate_reporting=DataErrorMethod.ERROR, i
             raise AugurError(duplicates_message)
         else:
             raise AugurError(f"Encountered unhandled duplicate reporting method: {duplicate_reporting!r}")
+
+
+def read_metadata_with_sequences(metadata, fasta, seq_id_column, seq_field='sequence'):
+    """
+    Read rows from *metadata* file and yield each row as a single dict that has
+    been updated with their corresponding sequence from the *fasta* file.
+    Matches the metadata record with sequences using the sequence id provided
+    in the *seq_id_column*. To ensure that the sequences can be matched with
+    the metadata, the FASTA headers should only contain the sequence id.
+
+    Note that metadata and sequence records that do not have a matching record
+    are skipped.
+
+    Reads the *fasta* file with `pyfastx.Fasta`, which creates an index for
+    the file to allow random access of sequences via the sequence id.
+    See pyfastx docs for more details:
+    https://pyfastx.readthedocs.io/en/latest/usage.html#fasta
+
+    Parameters
+    ----------
+    metadata: str
+        Path to a CSV or TSV metadata file
+
+    fasta: str
+        Path to a plain or gzipped FASTA file
+
+    seq_id_column: str
+        The column in the metadata file that contains the sequence id for
+        matching sequences
+
+    seq_field: str, optional
+        The field name to use for the sequence in the updated record
+
+    Yields
+    ------
+    dict
+        The parsed metadata record with the sequence
+    """
+    sequences = pyfastx.Fasta(fasta)
```

Comment:
Unfortunately, I don't think there's a way to manage this in pyfastx (at least not that I can find).

From our side, I guess the easiest way to DTRT would be to check the last modification time of the FASTA and *.fxi files and delete the *.fxi file if the FASTA was modified after the *.fxi file. This could potentially lead to unnecessary rebuilds of the index, but that's better than potentially stale caches.

Reply:
Did you benchmark how long it takes to build a `.fxi`? If it's not slow, we could force it every time and not worry about staleness. Otherwise, mtime checks are probably ok most of the time (but will almost certainly bite us in at least a few instances over time).

Reply:
I haven't done by own benchmarks, but I looked at the indexing benchmarks in the [Pyfastx paper](https://doi.org/10.1093/bib/bbaa368). Even with 4.25 million sequences for 27 billion bp genomes, the index building only took 39 seconds. 

I think it would be fine to delete the index and force rebuild every time by default.
We can also add the flag `--no-rebuild-index` to prevent rebuilding index if desired.